packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesModule.kt

```diff
@@ -237,15 +257,39 @@ class UpdatesModule(
                 return
               }
 
+              var shouldLaunch = false
               if (updatesServiceLocal.selectionPolicy.shouldLoadNewUpdate(
                   updateManifest.updateEntity,
                   launchedUpdate,
                   updateResponse.responseHeaderData?.manifestFilters
                 )
               ) {
+                // If "update" has failed to launch previously, then
+                // "launchedUpdate" will be an earlier update, and the test above
+                // will return true (incorrectly).
+                // We check to see if the new update is already in the DB, and if so,
+                // only allow the update if it has had no launch failures.
+                shouldLaunch = true
+                updateManifest.updateEntity?.let { updateEntity ->
+                  val storedUpdateEntity = updatesServiceLocal.databaseHolder.database.updateDao().loadUpdateWithId(
+                    updateEntity.id
+                  )
+                  updatesServiceLocal.databaseHolder.releaseDatabase()
+                  storedUpdateEntity?.let { storedUpdateEntity ->
+                    shouldLaunch = storedUpdateEntity.failedLaunchCount == 0
+                    logger.info("Stored update found: ID = ${updateEntity.id}, failureCount = ${storedUpdateEntity.failedLaunchCount}")
+                  }
+                }
+              }
+              if (shouldLaunch) {
```

Comment:
At this point, `shouldLaunch` can be false due to either `shouldLoadNewUpdate` returning false or `storedUpdateEntity.failedLaunchCount != 0`. But we only return `LoaderTask.RemoteCheckResultNotAvailableReason.UPDATE_PREVIOUSLY_FAILED`. We should differentiate between the two and return `UPDATE_REJECTED_BY_SELECTION_POLICY` when applicable.